src/fcntl.rs

```diff
@@ -596,39 +597,125 @@ pub fn fcntl(fd: RawFd, arg: FcntlArg) -> Result<c_int> {
     Errno::result(res)
 }
 
-// TODO: convert to libc_enum
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
 #[non_exhaustive]
 pub enum FlockArg {
     LockShared,
     LockExclusive,
-    Unlock,
     LockSharedNonblock,
     LockExclusiveNonblock,
-    UnlockNonblock,
 }
 
+/// Represents valid types for flock.
+///
+/// # Safety
+/// Types implementing this must be `!Clone`.
 #[cfg(not(any(target_os = "redox", target_os = "solaris")))]
-pub fn flock(fd: RawFd, arg: FlockArg) -> Result<()> {
-    use self::FlockArg::*;
+pub unsafe trait Flockable: AsRawFd {}
 
-    let res = unsafe {
-        match arg {
-            LockShared => libc::flock(fd, libc::LOCK_SH),
-            LockExclusive => libc::flock(fd, libc::LOCK_EX),
-            Unlock => libc::flock(fd, libc::LOCK_UN),
-            LockSharedNonblock => {
-                libc::flock(fd, libc::LOCK_SH | libc::LOCK_NB)
-            }
-            LockExclusiveNonblock => {
-                libc::flock(fd, libc::LOCK_EX | libc::LOCK_NB)
-            }
-            UnlockNonblock => libc::flock(fd, libc::LOCK_UN | libc::LOCK_NB),
+/// Represents an owned flock, which unlocks on drop.
+///
+/// See flock(2) for details on locking semantics.
+// `ManuallyDrop` is necessary to circumvent move out of `Drop` type error.
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+#[derive(Debug)]
+pub struct Flock<T: Flockable>(Option<T>);
+
+#[cfg(not(any(target_os = "redox", target_os = "solaris")))]
+impl<T: Flockable> Drop for Flock<T> {
+    fn drop(&mut self) {
+       if let Some(ref t) = self.0 {
+           // Result is ignored because flock has no documented failure cases.
```

Comment:
What do you mean?  Several failure cases are described in the man page.  And there's always a chance of an undocumented error.

Reply:
I meant that for unlocking specifically, none of the errors definitively seem to apply to unlock. Regardless, if we can't ignore the error then we can't implement `Drop`, which is the entire point of this effort. Further, flocks are "associated with an open file table entry" [^1], and since the inner object is dropped (when present), any remaining flocks will drop along with it.

If we want `Flock::unlock()` not ignore it, that's fine by me.

[^1]: [flock(2)](https://linux.die.net/man/2/flock)

Reply:
EBADF still applies.  I think we should panic on error (unless `std::thread::panicing == true`).